Elasticity inversion using NEML2

dewenyushu opened the pull request:
Inverse optimization for material properties leveraging neml2 stress derivatives.

dewenyushu commented on modules/tensor_mechanics/src/neml2/userobjects/CauchyStressFromNEML2UO.C:
```diff
@@ -45,11 +57,49 @@ CauchyStressFromNEML2UO::CauchyStressFromNEML2UO(const InputParameters & params)
+  }
+
+  // Set the parameter value using batch material from MOOSE
+  if (_reset_parameters)
+  {
+    // TODO: extend to multiple parameters
+    if (!_material_param_uo->outputReady())
+      mooseError("Batch material parameter is not ready for output");
+
+    auto model_param = model().named_parameters(true)[_reset_parameter_names[0]];
+    auto param_values = NEML2Utils::toNEML2Batched(_material_param_uo->getOutputData());
+
+    model_param = model_param.batch_expand_copy(param_values.batch_sizes());
+
+    model_param.copy_(param_values);
+  }
+}
```
@hugary1995 Thanks for your recent updates in neml2 which makes the `batch_expand_copy` to expand the scalar to proper batch size without needing to hard-code it in the input file. 

However, I still find that the parameter is not correctly set and/or consumed in neml2 material model. Therefore, would you mind helping me check if this is the correct way of setting material properties? Or am I not calling `precompute()` in the proper location?

dewenyushu replied:
> Can this wait a few days as I wrap up the current (giant) NEML2 refactor?

No prob at all - thanks for considering this in your current neml2 refactor.

dewenyushu commented on modules/combined/test/tests/material_inversion/test_param_set/model_test.i:
```diff
@@ -0,0 +1,169 @@
+    type = OptimizationData
+    variable = ux
+  []
+  [parametrization]
+    type = ConstantReporter
+    real_vector_names = 'coordx coordy youngs_modulus'
+    real_vector_values = '0 1 2; 0 1 2; 5 4 3'
+  []
+[]
```
@hugary1995 here is where I want to change the `youngs_modulus`. The idea is a bit complicated: 
 - set/transfer the `youngs_modulus` values at three physical locations in this reporter
 - extrapolate it in the function using `NearestReporterCoordinatesFunction`
 - transform the function to a material property `E_material` 
 - transform the material property as a batch material UO `E_batch_material`
 - then the UO is consumed in the `CauchyStressFromNEML2UO` for updating the neml2 material model parameter
 
So by changing the values here in the reporter, ideally we should see different displacement values. But I do not see the values changing.. I have checked that the values in `E_batch_material` changes correspondingly as I change the value in this reporter, which indicates an issue in setting the parameters in the `CauchyStressFromNEML2UO`

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2084567/) on 1140930 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/26750/site)

This comment will be updated on new commits.

dewenyushu committed:
Have solid_mechanics dependent on optimization

dewenyushu committed:
Get elasticity inversion to work using NEML2 derivatives.
Add inputs for an example using NEML2 derivatives and an example using hand-coded derivatives.
Clean up unused objects.

Co-development with Gary Hu.

Refs #26920

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2105159/) on ec9e418 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/26750/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/26750/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format b736baf4ce3e7ea35ed5f500c55af02bd31e7334`

dewenyushu committed:
Add neml2 requirement in the tests file, fix documentation, makefile

grmnptr committed:
Mpdify solid mechanics object registry.

dewenyushu commented:
I am confused about the documentation error and the phase_field & contact compilation error from civet. I cannot reproduce them on my local machine. Any ideas? Maybe @dschwen ?

dewenyushu committed:
Move BatchPropertyDerivative to framework and AdjointStrainStressGradNEML2InnerProduct to optimization module

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2122168/) on a5c3c2b wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>7f9748</th>
      <th colspan=4>#26750 a5c3c2</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.29%</td>
      <td>85.30%</td>
      <td>+0.01%</td>
      <td>73.33%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>101255</td>
      <td>101282</td>
      <td>+27</td>
      <td>11</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>17460</td>
      <td>17448</td>
      <td>-12</td>
      <td>4</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/26750/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/26750/coverage/framework)


## Modules coverage

### Optimization

<table>
  <thead>
    <tr>
      <th></th>
      <th>7f9748</th>
      <th colspan=4>#26750 a5c3c2</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>89.16%</td>
      <td>88.35%</td>
      <td>-0.81%</td>
      <td>0.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>1957</td>
      <td>1957</td>
      <td>-</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>238</td>
      <td>258</td>
      <td>+20</td>
      <td>20</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/26750/coverage_diff/optimization)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/26750/coverage/optimization)

### Solid mechanics

<table>
  <thead>
    <tr>
      <th></th>
      <th>7f9748</th>
      <th colspan=4>#26750 a5c3c2</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.78%</td>
      <td>84.77%</td>
      <td>-0.01%</td>
      <td>0.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>27113</td>
      <td>27113</td>
      <td>-</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>4868</td>
      <td>4873</td>
      <td>+5</td>
      <td>5</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/26750/coverage_diff/solid_mechanics)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/26750/coverage/solid_mechanics)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/26750/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/26750/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/26750/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/26750/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/26750/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/26750/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/26750/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/26750/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/26750/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/26750/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/26750/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/26750/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/26750/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/26750/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/26750/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/26750/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/26750/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/26750/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/26750/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/26750/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/26750/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/26750/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/26750/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/26750/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/26750/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/26750/coverage/stochastic_tools)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/26750/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/26750/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 73.33% is less than the suggested 90.0%
- `optimization` new line coverage rate 0.00% is less than the suggested 90.0%
- `solid_mechanics` new line coverage rate 0.00% is less than the suggested 90.0%



This comment will be updated on new commits.

dewenyushu committed:
Add test for BatchPropertyDerivative

dewenyushu committed:
Add test for AdjointStrainBatchStressGradInnerProduct

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2128143/) on 79fe3b0 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/26750/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/26750/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 7003f6aa88da6e901371f3ab8485044bd272bb26`

dewenyushu committed:
Set _output_ready=false in BatchMaterial's constructor

moosebuild commented:
Job [Docker moose-dev](https://civet.inl.gov/job/2130274/) on cab1af5 : invalidated by @dewenyushu

The failed test does not seem to be caused by this PR

hugary1995 commented:
Let me know if/when you want me to review.

hugary1995 commented on modules/combined/test/tests/optimization/invOpt_elasticity/forward.i:
```diff
@@ -0,0 +1,193 @@
+        formulation = TOTAL
+        incremental = true
+        volumetric_locking_correction = false
+        # generate_output = 'cauchy_stress_xx'
```
```suggestion
```

hugary1995 commented on modules/combined/test/tests/optimization/invOpt_elasticity/forward.i:
```diff
@@ -0,0 +1,193 @@
+[]
+
+[AuxVariables]
+  [dummy]
+  []
```
What's the use of this dummy variable?

dewenyushu replied:
Good catch. It is actually not used in here.

hugary1995 commented on modules/combined/test/tests/optimization/invOpt_elasticity/grad.i:
```diff
@@ -0,0 +1,187 @@
+        formulation = TOTAL
+        incremental = true
+        volumetric_locking_correction = false
+        # generate_output = 'cauchy_stress_xx'
```
```suggestion
```

hugary1995 commented on modules/combined/test/tests/optimization/invOpt_elasticity/grad.i:
```diff
@@ -0,0 +1,187 @@
+    prop_values = E
+  []
+[]
+[Functions]
```
```suggestion

[Functions]
```

hugary1995 commented on modules/combined/test/tests/optimization/invOpt_elasticity/grad.i:
```diff
@@ -0,0 +1,187 @@
+    value_name = parametrization/youngs_modulus
+  []
+[]
+[Reporters]
```
```suggestion

[Reporters]
```

hugary1995 commented on modules/combined/test/tests/optimization/invOpt_elasticity/grad.i:
```diff
@@ -0,0 +1,187 @@
+  # we do not compute them together as this is overwritting DiracKernel's residual calculation, which should be improved
+  residual_and_jacobian_together = false
+[]
+[Outputs]
```
```suggestion

[Outputs]
```

hugary1995 commented on modules/combined/test/tests/optimization/invOpt_elasticity/grad.i:
```diff
@@ -0,0 +1,187 @@
+  solve_type = NEWTON
+  petsc_options_iname = '-pc_type'
+  petsc_options_value = 'lu'
+  # we do not compute them together as this is overwritting DiracKernel's residual calculation, which should be improved
```
Could you open a ticket about this issue? We could use the new `mooseDocumentedError` API on this.

dewenyushu replied:
Ref issue #27258

hugary1995 commented on modules/optimization/doc/content/source/vectorpostprocessors/AdjointStrainBatchStressGradInnerProduct.md:
```diff
@@ -0,0 +1,7 @@
+# AdjointStrainBatchStressGradInnerProduct
+
+This component is designed to compute the gradient of the objective function concerning specific properties. It achieves this by computing the inner product of the property derivative and the strain resulting from the forward simulation. The property derivative is extracted from a batch material of the [BatchPropertyDerivative](BatchPropertyDerivative.md) type.
```
@lynnmunday If this is all you need, we could provide this 100 times more efficiently... We should probably talk about this.

NEML2 can give you this information in two ways:
1. Compute derivative of stress w.r.t. parameter, then compute its action on the adjoint strain.
2. Directly compute the action of the parameter derivative on the adjoint strain.

We are doing 1 right now. 2 is more efficient.

hugary1995 replied:
Oh to be clear, we don't need to address this in this PR.

dewenyushu replied:
Noted in issue #26920

hugary1995 reviewed (approved):
Very exciting new feature! I just a few minor suggestions and a few things to discuss.

dewenyushu committed:
Update framework/doc/content/source/userobjects/BatchPropertyDerivative.md

Co-authored-by: Gary (Tianchen) Hu <thu@anl.gov>

dewenyushu committed:
Apply suggestions from code review

Co-authored-by: Gary (Tianchen) Hu <thu@anl.gov>

dewenyushu committed:
Address comments

moosebuild commented:
Job [Conda MOOSE (Linux)](https://civet.inl.gov/job/2153487/) on 574d6d1 : invalidated by @dewenyushu

The failed test does not seem to be caused by this PR

hugary1995 commented on framework/include/userobjects/BatchPropertyDerivative.h:
```diff
@@ -0,0 +1,34 @@
+  void batchCompute() override {}
+};
+
+typedef BatchPropertyDerivative<RankTwoTensor, Real> BatchPropertyDerivativeRankTwoTensorReal;
```
Uhh, idk, maybe this is fine. I guess we are going to wrap everything inside an Action in the end, so this isn't that big of a concern. I am marking this as resolved now.

dschwen commented:
This looks fine to me. I'm thinking about changes to batch material to allow runtime selection of gathered properties/variables etc. I'll have to do some profiling to see if all that template meta-programming hell was worth the effort, or if I can throw some of it out in favor of runtime code. But that shouldn't hold up this PR
